### ISSUE TITLE
[lldp] LLDP does not have logic to gracefully handle port deletion

### ISSUE THREAD

**Issue by alexrallen** [Azure/sonic-buildimage#9802]
#### Description

When we delete a port using dynamic port breakout lldp does not gracefully handle the port deletion.

This is likely due to the fact that in the lldpmgrd logic it attempts to lookup the alias and port name of a deleted port in Config DB, a request that we know will fail.

https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-lldp/lldpmgrd#L288

#### Steps to reproduce the issue:
1. Use dynamic port breakout to remove a port running on the switch (go from a 2x to 1x breakout mode or similar)

#### Describe the results you received:

```
Jan 19 15:50:40.908722 r-lionfish-14 ERR lldp#lldpmgrd[33]: Port 'Ethernet236' not found in PORT table in Config DB. Using port name instead of port alias.
```

This error is seen for all removed ports. 

#### Describe the results you expected:

No Error.

**Comment by dgsudharsan** [comment]
@zhenggen-xu  @praveen-li  Can you please investigate?

**Comment by praveen-li** [comment]
Kindly share initial config and command used for breakout. 
Seems like config_Db did not have the entry initially for a port and that port is added later by breakout. But lldrmgrd complains that port is not present in config db which is a local copy taken at start.

**Comment by alexrallen** [comment]
@praveen-li I have attached a techsupport which should give you access to the state of config_db.json as well as the state of CONFIG_DB. 

What we did was breakout 4 ports from 100G to 4x25G and then clean  up those breakouts. You should be able to follow the syslog and see that pattern. 

[sonic_dump_r-lionfish-14_20220119_161453.tar.gz](https://github.com/Azure/sonic-buildimage/files/7943734/sonic_dump_r-lionfish-14_20220119_161453.tar.gz)

**Comment by liat-grozovik** [comment]
@praveen-li this issue is for 202111. can you please help to provide ETA for fix to be available?

**Comment by praveen-li** [comment]
As per Code, LLDP listens to DEL event in config DB. So we need more data on:

-- Do you observe LLDP listens to only specific DEL event in config DB, but not all.
-- If listens to all, then Is it a regression introduced later in LLDP or do we surely know, it is day 1 issue which is exposed by DPB?


Thanks,

Regards
Praveen
________________________________

**Comment by alexrallen** [comment]
Wouldn't this line of code fail when a port is deleted?
https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-lldp/lldpmgrd#L299

It attempts to pull a port that no longer exists from CONFIG_DB here
https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-lldp/lldpmgrd#L140

**Comment by praveen-li** [comment]
When DPB deletes the Port, Config Changes is
```
{
'PORT': {
'Ethernet8': None
}
}
```

Here we seems to be checking >>>> "if ("alias" in fvp_dict or "description" in fvp_dict)" in https://github.com/Azure/sonic-buildimage/blob/d112e7cca9087d02892ada9d70bddcc5d5616c24/dockers/docker-lldp/lldpmgrd#L297

BTW, we do not have a local repro for this in LinkedIn, so if possible please log all FVs in case of config_db changes during DPB. FVs data may give us some idea as why LLDP is failing here, or is it a timing issue.


Thanks,

Regards
Praveen
________________________________

**Comment by alexrallen** [comment]
Hrm okay looking at this a bit more I think you are correct @praveen-li that this is a timing issue.

Specifically I believe this is a race condition between APP_DB and LLDP. 

In the log we can see directly before portmgrd does a deletion of the port it sets it to admin down. 

```
Jan 19 15:50:40.907150 r-lionfish-14 NOTICE swss#portmgrd: :- doTask: Configure Ethernet236 admin status to down
Jan 19 15:50:40.907150 r-lionfish-14 NOTICE swss#portmgrd: :- doTask: Delete Port: Ethernet236
```

I believe that before LLDP handles the admin_down operation the port is also deleted so that when it gets to the following line...
https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-lldp/lldpmgrd#L319

It will have passed the `if "oper_status" in fvp_dict` check on line 317 and then results in the error message on line 319. 

Granted, I am not seeing reproductions any more manually, I am checking internally to see if we are seeing reproductions of this error message, due to the apparent racing nature of it, they might be infrequent.

**Comment by zhenggen-xu** [comment]
Is this a temp issue and would be recovered by it own later? 

By looking at the code, it does have timing issue where the code use different snapshot of the app PORT table make decision vs execution.  If this is the case, we may need refactor the code to take a snapshot of the PORT table once then make decision and execution based on that.

**Comment by alexrallen** [comment]
Looks like it does recover yes, the error log is seen only during the race condition.

**Comment by zhenggen-xu** [comment]
I agree, we should fix it.  The appDB PORT table should be read once to local memory in the while loop, and then it should be used in the functions in the loop. This can avoid inconsistency in different functions having potential different views of the table.
